File: leetcode/387-first-unique-character-in-a-string/main.go
package main

import "fmt"

/*
	naive solution
	hashtable
	beats 10%
*/
func firstUniqChar(s string) int {
	hash := make(map[byte]int)
	for i := 0; i < len(s); i++ {
		if _, ex := hash[s[i]]; !ex {
			hash[s[i]] = 1
		} else {
			hash[s[i]]++
		}
	}
	for i := 0; i < len(s); i++ {
		if v, _ := hash[s[i]]; v == 1 {
			return i
		}
	}
	return -1
}

// optimized solutin for the string that contains only a-z
// array
func firstUniqChar1(s string) int {
	history := [26]int{}
	for i := 0; i < len(s); i++ {
		history[s[i]-'a']++
	}
	for i := 0; i < len(s); i++ {
		if history[s[i]-'a'] == 1 {
			return i
		}
	}
	return -1
}

func main() {
	fmt.Println(firstUniqChar("loveleetcode"))
	fmt.Println(firstUniqChar("https://leetcode.com/problems/first-unique-character-in-a-string/"))
	fmt.Println(firstUniqChar1("loveleetcode"))
	fmt.Println(firstUniqChar1("irstuniquecharacterinastring"))
}

File: leetcode/387-first-unique-character-in-a-string/main.py
"""
    1st: hashtable
	  136ms beats 57.31%
"""


class Solution(object):
    def firstUniqChar(self, s):
        """
        :type s: str
        :rtype: int
        """
        ht = {}
        for c in s:
            if c not in ht:
                ht[c] = 1
            else:
                ht[c] += 1
        for i in range(len(s)):
            if ht[s[i]] == 1:
                return i
        return -1
